Move common autograd utils functions from gen_variable_type.py to api/autograd.py.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53343 Implement public API InferenceMode and its error handling
* #53342 Move as_view/increment_version to its separate key.
* #53341 Move view handling logic to gen_inplace_or_view_type.py
* **#53340 Move common autograd utils functions from gen_variable_type.py to api/autograd.py.**

Differential Revision: [D26973914](https://our.internmc.facebook.com/intern/diff/D26973914)

Commit:
Move common autograd utils functions from gen_variable_type.py to api/autograd.py.

[ghstack-poisoned]

Commit:
Update on "Move common autograd utils functions from gen_variable_type.py to api/autograd.py."




[ghstack-poisoned]

Commit:
Update on "Move common autograd utils functions from gen_variable_type.py to api/autograd.py."




[ghstack-poisoned]

Commit:
Update on "Move common autograd utils functions from gen_variable_type.py to api/autograd.py."




[ghstack-poisoned]

ailzhang commented:
Test Plan: verified the generated VariableTypeEverything.cpp are the same after this PR.

albanD commented on tools/autograd/gen_autograd.py:
```diff
@@ -23,7 +23,15 @@
 
 import argparse
 import os
+# Favoring from...to instead of import...as
```
Why this comment here?
I understand the change in the other files, but I'm not sure having just one comment here is useful?

ailzhang replied:
Yea I was just thinking adding a note somewhere why we changed it but couldn't find a good place. Actually I will just remove this. Few people will wonder about from...to I think :D

albanD commented on tools/autograd/gen_autograd.py:
```diff
@@ -23,7 +23,15 @@
+# due to a discrepancy between them up to Python version 3.6
+# https://stackoverflow.com/questions/24807434/imports-in-init-py-and-import-as-statement
+from tools.codegen.api import cpp
+from tools.codegen.api.autograd import *
```
nit: I am not a fan of `*` here, can you name explicitly the objects you actually use?

albanD reviewed (approved):
LGTM

Commit:
Update on "Move common autograd utils functions from gen_variable_type.py to api/autograd.py."




[ghstack-poisoned]

bhosmer commented on tools/codegen/api/autograd.py:
```diff
@@ -120,3 +120,94 @@ class DifferentiableOutput:
+    func: NativeFunction
+    info: Optional[DifferentiabilityInfo]
+
+def dispatch_strategy(fn: NativeFunctionWithDifferentiabilityInfo) -> str:
```
Since we're moving this, it might be the right time to strip the old comments (they're still from the world of Tensors and Variables).

I'm actually not even sure the right way to describe this function's role in the current system, but it's probably out of scope for this PR to, like, recontexualize it. Maybe just a TODO above the comment so people reading don't try to make sense of it w.r.t. the current system... before this PR they could at least look at the history and see that it's like 3 years old, but afterwards there's nothing to tell them it's not fresh code.

ailzhang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -38,6 +38,7 @@
 from .gen_trace_type import should_trace
 
 from tools.codegen.code_template import CodeTemplate
+from tools.codegen.api import cpp
```
Yea if this line is removed mypy complains about cpp. 
Likely it was imported implicitly using `*` so after my change it's required.

Commit:
Update on "Move common autograd utils functions from gen_variable_type.py to api/autograd.py."




[ghstack-poisoned]

Commit:
Update on "Move common autograd utils functions from gen_variable_type.py to api/autograd.py."


Differential Revision: [D26973914](https://our.internmc.facebook.com/intern/diff/D26973914)

[ghstack-poisoned]
